components/feature/qr/src/main/java/mozilla/components/feature/qr/CustomOverlay.kt

```diff
@@ -0,0 +1,102 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.qr
+
+import android.app.Activity
+import android.content.Context
+import android.graphics.Bitmap
+import android.graphics.Canvas
+import android.graphics.Color
+import android.graphics.Paint
+import android.graphics.PorterDuff
+import android.graphics.PorterDuffXfermode
+import android.graphics.RectF
+import android.graphics.Rect
+import android.util.AttributeSet
+import android.util.DisplayMetrics
+import android.widget.ImageView
+import android.widget.LinearLayout
+
+/**
+ * A custom overlay [View] that sets the opacity of the background to dim it.
+ * This gives a focus like look to the qr reticle.
+ * @property windowFrame a reference to a [Bitmap], used to create the window frame.
+ * @property activity a reference to an [Activity], used to refer the child views of
+ * calling parent view.
+ * @property ALPHA defines the opacity of the background.
+ * @property RADIUS defines the radius of the rectangle corners.
+ */
+
+class CustomOverlay @JvmOverloads constructor (
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyleAttr: Int = 0
+) : LinearLayout(context, attrs, defStyleAttr) {
+
+    private var windowFrame: Bitmap? = null
+    private val activity = context as Activity
+
+    companion object {
+        const val RADIUS = 30.0f
+        const val ALPHA = 170
+    }
+
+    /**
+     * This is called before the canvas is drawn. Initializes the windowFrame only once
+     * @param canvas The canvas to draw upon.
+     */
+    override fun dispatchDraw(canvas: Canvas) {
+        super.dispatchDraw(canvas)
+
+        windowFrame ?: createWindowFrame()
+
+        canvas.drawBitmap(windowFrame!!, 0f, 0f, null)
+    }
+
+    override fun isEnabled() = false
+
+    override fun isClickable() = false
+
+    /**
+     * This function creates the windows frame around our R.id.qr_reticle ImageView
+     * dynamically, so this will scale for all display sizes. Sets an alpha of 170%
+     * to give a dimmed look for the background
+     */
+    private fun createWindowFrame() {
+        val dm = DisplayMetrics()
+        val imageView = activity.findViewById(R.id.qr_reticle) as ImageView
```

Comment:
It's a bit fragile that this view looks for other views in its parent activity. I wonder if that implementation could be unified: One ImageView that draws the reticle and the frame (by extending `AppCompatImageView`)?

Reply:
Actually I had thought of this, but instead using a ``LinearLayout``. I was going to have child view that would be an ``ImageView`` and then get the reference of its dimensions and draw this overlay. Turns out child views are rendered at a later stage. Hence I had to use the view in its parent, as it would have been drawn already there. Also extending a simple ``View`` instead of ``LinearLayout`` or ``AppCompatImageView`` would have done the work as we have to set the parent view, i.e the overlay to span over the whole screen i.e. ``match_parent``. So no use of ``AppCompatImageView`` there because we are not setting any *drawable* for it.  The child of these views would've been an ImageView which would contain the *reticle*. I hope this clears the reason?